Compose Destinations library changes

raamcosta opened the pull request:
Hi Philipp!

I'm the author of [Compose Destinations library](https://github.com/raamcosta/compose-destinations) and a fan of your work (also a paid subscriber on youtube).

I was thinking about testing my library in a bigger app for a while and finally got around to doing it with your app. 
I'll leave the changes here in a PR, so you can take a look at them. 

I am 100% sure that if you take a good look at them, you will see the great benefits of the library. 
I would be thrilled if you could do that and give me feedback and/or ask questions etc.
Then.. who knows maybe you actually like the idea of using it in the project 🙂 If not, then it was good fun and served the purpose to test my library features.

With this library in place, it would be super easy to later incorporate animations when navigating between screens, and you get all the type-safety and less boilerplate (AND you get to pass Parcelables/Enums between screens if you want!)

I'll be paying attention to the PR for comments, or hit me up on [Twitter DMs](https://twitter.com/raamcosta).

Thank you in advance for your time!

Rafael Costa

philipplackner commented:
Hi Rafael,

thanks for the PR, looks very promising. Since I'll be streaming in an hour and don't have time now to review it, I'll check it for the next stream which will be next year :)

All the best,
Philipp

raamcosta commented on build.gradle:
```diff
@@ -11,7 +11,7 @@ buildscript {
     dependencies {
         classpath "com.android.tools.build:gradle:7.0.2"
-        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.21"
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.10"
```
I updated kotlin to 1.6.10 (latest) because I had this issue before:

https://youtrack.jetbrains.com/issue/KT-45545

Which I'm sure has nothing to do with my library / changes 🤷 
Anyway, it should be the same for you, but let me know if you find any issues. 
Also updated compose to match the new kotlin version.

raamcosta commented on app/src/main/java/com/plcoding/socialnetworktwitch/feature_profile/presentation/search/SearchScreen.kt:
```diff
@@ -17,24 +17,22 @@ import androidx.hilt.navigation.compose.hiltViewModel
 
-@ExperimentalCoilApi
-@ExperimentalMaterialApi
+@OptIn(ExperimentalMaterialApi::class, ExperimentalCoilApi::class)
```
OptIn's are preferable to normal @Experimental annotations on @Destination annotated Screens because otherwise the generated Destination which calls this Composable needs to be marked @Experimental, and then the NavGraphs object also does and so on..

raamcosta commented on app/build.gradle:
```diff
@@ -50,6 +61,14 @@ android {
+                '-Xopt-in=kotlin.RequiresOptIn',
+        ]
+    }
+}
```
Needed to use @OptIn with no warnings.

raamcosta commented on app/build.gradle:
```diff
@@ -4,6 +4,18 @@ plugins {
+        release {
+            kotlin.srcDir("build/generated/ksp/release/kotlin")
+        }
+    }
```
Needed to make the IDE look at the generated files from KSP. Known KSP "issue" that we hopefully won't need once they improve it.

raamcosta commented on app/src/main/java/com/plcoding/socialnetworktwitch/feature_chat/presentation/message/MessageScreen.kt:
```diff
@@ -29,24 +29,30 @@ import com.plcoding.socialnetworktwitch.core.presentation.ui.theme.ProfilePictur
+    val remoteUsername: String,
+    val encodedRemoteUserProfilePictureUrl: String,
+    val chatId: String?
+)
```
I left these `*NavArgs` classes right beside the screen composables just so you can more easily check them, but I'd move them to their own file 🙂

raamcosta commented:
Fixed conflicts with your latest changes and updated compose destinations to the latest release.

Actually found a small bug with a 1.1.0-beta new feature when I tried it in your app, and so I had to release a new 1.1.1-beta release 😄 
I'm glad I used your app to test it! ❤️

raamcosta commented on app/build.gradle:
```diff
@@ -40,8 +52,7 @@ android {
     composeOptions {
-        kotlinCompilerExtensionVersion compose_version
-        kotlinCompilerVersion '1.5.10'
+        kotlinCompilerExtensionVersion "1.1.0-rc02"
```
The current version of compose is weird in that we need to use a different compiler version than the one used in the actual dependencies.

Next compose versions might go back to usual, so you could go back to `compose_version` if that's the case.

raamcosta commented:
Solved conflicts of your last stream and updated Compose Destinations to latest 🙂
